salt/states/file.py

```diff
@@ -557,7 +557,18 @@ def process_symlinks(filenames, symlinks):
             managed_directories.add(mdest)
             keep.add(mdest)
 
-    return managed_files, managed_directories, managed_symlinks, keep
+    # Sets are randomly ordered. We need to make sure symlinks are always at the
+    # end, so we need to use a list
+    new_managed_files = list(managed_files)
+    # Now let's move all the symlinks to the end
+    for symlink in managed_symlinks:
+        for file in managed_files:
+            if file[0].endswith(symlink[0]):
```

Comment:
The problem is that the ``symlink`` contains only the name of the file and not its entire path. Adding an ``os.sep`` solves the scenario you describe. But another scenario occurs to me where the symlink share its name with a file in another directory that is not a symlink. But as you said, that may not be an issue since it is just moving it to the end of the list... unless that file/directory is the target for the symlink we care about.